crates/wasm-smith/src/lib.rs

```diff
@@ -2255,20 +2249,46 @@ impl Module {
         max_minimum: u64,
         max_required: bool,
     ) -> Result<(u64, Option<u64>)> {
-        let min = u.int_in_range(0..=max_minimum)?;
+        let min = gradually_grow(u, 0, max_minimum)?;
         let max = if max_required || u.arbitrary().unwrap_or(false) {
-            Some(if min == max_minimum {
-                max_minimum
-            } else {
-                u.int_in_range(min..=max_minimum)?
-            })
+            Some(u.int_in_range(min..=max_minimum)?)
         } else {
             None
         };
         Ok((min, max))
     }
 }
 
+/// This function generates a number between `min` and `max`, favoring values
+/// closer to `min`.
+///
+/// The thinking behind this function is that it's used for things like offsets
+/// and minimum sizes which, when very large, can trivially make the wasm oom or
+/// abort with a trap. This isn't the most interesting thing to do so it tries
+/// to favor numbers closer to `min` if possible.
+fn gradually_grow(u: &mut Unstructured, min: u64, max: u64) -> Result<u64> {
+    // The idea behind how this works is that we use the input `u` to determin
+    // the ceiling, ideally between min/max, and then generate a number between
+    // min and our ceiling.
+    //
+    // The actual numbers here may need some tuning over time, but at least
+    // locally when I threw a bunch of random data at this it would generate
+    // instances that would oom on instantiation 1% of the time. If this
+    // function changed to simply use `int_in_range(min, max)` it would oom 50%
+    // of the time.
+    assert!(min <= max);
+    let mut cur_max = min.saturating_add(10);
+    loop {
+        if cur_max >= max {
+            break u.int_in_range(min..=max);
+        }
+        if u.int_in_range(0..=4)? == 0 {
+            break u.int_in_range(min..=cur_max);
+        }
+        cur_max += cur_max / 2;
+    }
```

Comment:
Instead of this adhoc gradual growth, lets use a cubic easing function:

```rust
/// Map a value from within the input range linearly to the output range.
///
/// For example, mapping `0.5` from the input range `0.0..1.0` to the output
/// range `1.0..3.0` produces `2.0`.
fn map_linear(
    value: f64,
    Range {
        start: in_low,
        end: in_high,
    }: Range<f64>,
    Range {
        start: out_low,
        end: out_high,
    }: Range<f64>,
) -> f64 {
    assert!(!value.is_nan(), "{}", value);
    assert!(value.is_finite(), "{}", value);
    assert!(in_low < in_high, "{} < {}", in_low, in_high);
    assert!(out_low < out_high, "{} < {}", out_low, out_high);
    assert!(value >= in_low, "{} >= {}", value, in_low);
    assert!(value <= in_high, "{} <= {}", value, in_high);

    let dividend = out_high - out_low;
    let divisor = in_high - in_low;
    let slope = dividend / divisor;
    let result = out_low + (slope * (value - in_low));

    assert!(result >= out_low, "{} >= {}", result, out_low);
    assert!(result <= out_high, "{} <= {}", result, out_high);
    result
}

/// Map a value from within the input range cubically to the output range.
///
/// For example, mapping `0.333333` from the input range `0.0..1.0` to the
/// output range `0.0..100.0` will result in approximately `3.7`.
fn map_cubed(
    value: f64,
    Range {
        start: in_low,
        end: in_high,
    }: Range<f64>,
    Range {
        start: out_low,
        end: out_high,
    }: Range<f64>,
) -> f64 {
    assert!(!value.is_nan(), "{}", value);
    assert!(value.is_finite(), "{}", value);
    assert!(in_low < in_high, "{} < {}", in_low, in_high);
    assert!(out_low < out_high, "{} < {}", out_low, out_high);
    assert!(value >= in_low, "{} >= {}", value, in_low);
    assert!(value <= in_high, "{} <= {}", value, in_high);

    let x = map_linear(value, in_low..in_high, 0.0..1.0);
    let unscaled = x * x * x;
    let result = map_linear(unscaled, 0.0..1.0, out_low..out_high);

    assert!(result >= out_low, "{} >= {}", result, out_low);
    assert!(result <= out_high, "{} <= {}", result, out_high);
    result
}

fn gradually_grow(u: &mut Unstructured, min: u64, max: u64) -> Result<u64> {
    let min = f64::try_from(min).unwrap();
    let max = f64::try_from(max).unwrap();
    let x = u.arbitrary::<u32>()?;
    let x = f64::from(x);
    let x = map_cubed(x, f64::from(u32::MIN)..f64::from(u32::MAX), min..max);
    Ok(u64::try_from(x.round()).unwrap())
}
```

Given uniform inputs in the range 0..1000, this results in the following output distribution:

```
=== Inputs ===
# Number of samples = 16384
# Min = 0
# Max = 1000
#
# Mean = 502.8545532226551
# Standard deviation = 290.2145333477139
# Variance = 84224.47536623133
#
# Each ∎ is a count of 17
#
   0 ..   50 [ 834 ]: ∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎
  50 ..  100 [ 834 ]: ∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎
 100 ..  150 [ 772 ]: ∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎
 150 ..  200 [ 822 ]: ∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎
 200 ..  250 [ 761 ]: ∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎
 250 ..  300 [ 819 ]: ∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎
 300 ..  350 [ 841 ]: ∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎
 350 ..  400 [ 823 ]: ∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎
 400 ..  450 [ 821 ]: ∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎
 450 ..  500 [ 783 ]: ∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎
 500 ..  550 [ 818 ]: ∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎
 550 ..  600 [ 812 ]: ∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎
 600 ..  650 [ 826 ]: ∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎
 650 ..  700 [ 802 ]: ∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎
 700 ..  750 [ 832 ]: ∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎
 750 ..  800 [ 819 ]: ∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎
 800 ..  850 [ 805 ]: ∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎
 850 ..  900 [ 842 ]: ∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎
 900 ..  950 [ 840 ]: ∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎
 950 .. 1000 [ 878 ]: ∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎

=== Outputs ===
# Number of samples = 16384
# Min = 0
# Max = 1000
#
# Mean = 253.90454101562332
# Standard deviation = 286.36370736253974
# Variance = 82004.1728944183
#
# Each ∎ is a count of 119
#
   0 ..   50 [ 5967 ]: ∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎∎
  50 ..  100 [ 1552 ]: ∎∎∎∎∎∎∎∎∎∎∎∎∎
 100 ..  150 [ 1092 ]: ∎∎∎∎∎∎∎∎∎
 150 ..  200 [  885 ]: ∎∎∎∎∎∎∎
 200 ..  250 [  736 ]: ∎∎∎∎∎∎
 250 ..  300 [  666 ]: ∎∎∎∎∎
 300 ..  350 [  544 ]: ∎∎∎∎
 350 ..  400 [  548 ]: ∎∎∎∎
 400 ..  450 [  478 ]: ∎∎∎∎
 450 ..  500 [  456 ]: ∎∎∎
 500 ..  550 [  419 ]: ∎∎∎
 550 ..  600 [  388 ]: ∎∎∎
 600 ..  650 [  375 ]: ∎∎∎
 650 ..  700 [  350 ]: ∎∎
 700 ..  750 [  359 ]: ∎∎∎
 750 ..  800 [  346 ]: ∎∎
 800 ..  850 [  306 ]: ∎∎
 850 ..  900 [  294 ]: ∎∎
 900 ..  950 [  341 ]: ∎∎
 950 .. 1000 [  282 ]: ∎∎
```

This should give us roughly our desired distribution shape without consuming any additional bytes from the `Unstructured`.

Reply:
I'm glad at least one of us is aware of how to actually do this correctly.

Reply:
Ok so I implemented this and am getting some interesting results. My goal was to have oom-produced modules happen maybe 1-2% of the time, but doing the cubic thing here cases oom ~30% of the time still. In retrospect this makes sense. I'm considering >1gb to be oom which is 1/4 of the way through the range from the minimum to the maximum of 32-bit memories. Turns out the cube root of 1/4 is ~0.63, which roughly equates to 35% ish of the random choices (in the range of 0-1) being in the oom area.

Thinking more about this though I like the idea of what's going on here, basically generating a random float between 0 and 1 and then graphing some curve from 0-1, using the resulting value to scale up to the range we actually are interested in. I don't think y=x^3 is what we want to do here because at least for memories and oom that still biases 1gb+ memories too much.

According to wolfram alpha if we want 98% of the memories to be in-bounds (e.g. less than 1/4 in the 0-1 range) then the solution to `0.98^n = 1/4` is `n ~= 69`. While possible that seems a little silly to do a 69-th power of a function just to get the desired range.

I wonder if we can still do something similar here? What do you think of doing something like:

* Generate a 32-bit number
* Map it to 0-1
* If the number is less than 0.99, then generate a guaranteed-in-bounds size. This would be new parameters to the `gradually_grow` function where we'd just linearly map 0-0.99 into the output range known to be guaranteed inbounds.
* If the number is 0.99 or above we map the 0.99 to 1 range to the output range which is always out of bounds.

This would largely amount to basically having a function that looks like:

```
f(x) = scale(0..1, x^3, 0..MIN)       if x < 0.99
     = scale(0..1, x, MIN..MAX)     if x >= 0.99
```

where `MIN` is the maximum value that's still in-bounds and `MAX` is the absolute maximum value that we're allowed to generate. This is not a continuous function like x^3 (or x^69) but I think would suffice for our purposes, give us a parameter to tweak (the 0.99 threshold and the MIN), and still consume only the 32-bit integer from the input data?

Reply:
Ok I went ahead and pushed up what I was thinking